File: python_examples/demos/upload_image_ask.py
import asyncio
import os
from pathlib import Path
from dotenv import load_dotenv
from mix_python_sdk import Mix
from mix_python_sdk.helpers import stream_and_send


def upload_sample_image(mix, session_id: str) -> str:
    """Upload sample.jpg to the session"""
    # Get the path to sample_files directory relative to this script
    script_dir = Path(__file__).parent
    image_path = script_dir / "../../sample_files/sample.jpg"
    image_path = image_path.resolve()

    with open(image_path, "rb") as f:
        image_file_info = mix.files.upload_session_file(
            id=session_id,
            file={
                "file_name": "sample.jpg",
                "content": f,
                "content_type": "image/jpeg",
            },
        )
    print(f"✅ Uploaded File URL: {image_file_info.url}")
    return image_file_info.url


async def main():
    load_dotenv()
    api_key = os.getenv("OPENROUTER_API_KEY")
    if not api_key:
        raise ValueError("OPENROUTER_API_KEY not found in environment variables")

    async with Mix(server_url=os.getenv("MIX_SERVER_URL")) as mix:
        mix.system.get_health()
        # mix.authentication.store_api_key(api_key=api_key, provider="anthropic")
        mix.preferences.update_preferences(
            preferred_provider="anthropic",
            main_agent_model="claude-sonnet-4-5",
        )

        # session creation
        session = mix.sessions.create(title="Image Analysis Demo")

        # Upload sample image
        uploaded_file_url = upload_sample_image(mix, session.id)

        # Ask about the uploaded image
        user_msg = f"Explain {uploaded_file_url}"

        def handle_tool_complete(data):
            """Only show actual tool output content, skip completion messages"""
            # Only print if there's actual output content (not just completion status)
            if hasattr(data, "output") and data.output:
                print(f"\n{data.output}")
            elif hasattr(data, "result") and data.result:
                print(f"\n{data.result}")

        await stream_and_send(
            mix,
            session_id=session.id,
            message=user_msg,
            on_thinking=lambda text: print(text, end="", flush=True),
            on_content=lambda text: print(text, end="", flush=True),
            on_tool_execution_complete=handle_tool_complete,
            on_error=lambda error: print(f"\n❌ {error}"),
        )


if __name__ == "__main__":
    asyncio.run(main())

File: typescript_examples/demos/upload_image_ask.ts
#!/usr/bin/env tsx
/**
 * Image upload and analysis demo - upload image and ask questions about it.
 */

import { config } from "dotenv";
import { Mix } from "mix-typescript-sdk";
import { sendWithCallbacks } from "../helpers.js";
import { readFileSync } from "fs";
import { resolve } from "path";
import { fileURLToPath } from "url";
import { dirname } from "path";

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

config();

async function main() {
  const mix = new Mix({
    serverURL: process.env.MIX_SERVER_URL,
  });

  // Update preferences
  await mix.preferences.update({
    preferred_provider: "anthropic",
    main_agent_model: "claude-sonnet-4-5",
  });

  // Create session
  const session = await mix.sessions.create({
    title: "Image Analysis Demo",
  });

  // Upload image file
  const imagePath = resolve(__dirname, "../../sample_files/sample.jpg");
  const imageContent = readFileSync(imagePath);
  const imageFile = new File([imageContent], "sample.jpg", {
    type: "image/jpeg",
  });

  const fileInfo = await mix.files.upload({
    id: session.id,
    requestBody: {
      file: imageFile,
    },
  });

  console.log(`Uploaded: ${fileInfo.url}\n`);

  // Handle tool events
  function handleTool(tool: any) {
    if (tool.name !== "show_media" || !tool.input) return;

    try {
      const input = typeof tool.input === "string" ? JSON.parse(tool.input) : tool.input;
      const outputs = input.outputs || [];

      for (const output of outputs) {
        console.log(`\n${output.title}`);
        if (output.description) console.log(`   ${output.description}`);
        if (output.path) console.log(`   ${output.path}`);
        console.log();
      }
    } catch (e) {
      // Silently ignore parse errors
    }
  }

  // Ask about the image
  await sendWithCallbacks(
    mix,
    session.id,
    `Explain ${fileInfo.url}`,
    {
      onTool: handleTool,
      onContent: (text) => {
        // Add line breaks after periods before capital letters for readability
        const formatted = text.replace(/(\.)([A-Z])/g, "$1\n$2");
        process.stdout.write(formatted);
      },
      onError: (error) => console.log(`\n❌ ${error}`),
    }
  );
}

main()
  .catch(console.error)
  .finally(() => process.exit(0));
